src/mutable_data.rs

```diff
@@ -68,48 +127,537 @@ impl MutableData {
+    pub fn set_user_permissions(
+        &mut self,
+        user: PublicKey,
+        permissions: PermissionSet,
+        version: u64,
+        requester: PublicKey,
+    ) -> Result<(), DataError> {
+        if !self.is_action_allowed(requester, Action::ManagePermissions) {
+            return Err(DataError::AccessDenied);
+        }
+        if version != self.version + 1 {
+            return Err(DataError::InvalidSuccessor(self.version));
+        }
+        let _prev = self.permissions.insert(user, permissions);
+        self.version = version;
+        Ok(())
+    }
+
+    /// Delete permissions for the provided user.
+    pub fn del_user_permissions(
+        &mut self,
+        user: &PublicKey,
+        version: u64,
+        requester: PublicKey,
+    ) -> Result<(), DataError> {
+        if !self.is_action_allowed(requester, Action::ManagePermissions) {
+            return Err(DataError::AccessDenied);
+        }
+        if version != self.version + 1 {
+            return Err(DataError::InvalidSuccessor(self.version));
+        }
+        if !self.permissions.contains_key(user) {
+            return Err(DataError::NoSuchKey);
+        }
+        let _ = self.permissions.remove(user);
+        self.version = version;
+        Ok(())
+    }
+
+    /// Delete user permissions without performing any validation.
+    pub fn del_user_permissions_without_validation(
+        &mut self,
+        user: &PublicKey,
+        version: u64,
+    ) -> bool {
+        if version <= self.version {
+            return false;
+        }
+        let _ = self.permissions.remove(user);
+        self.version = version;
+        true
+    }
+
+    /// Change owner of the mutable data.
+    pub fn change_owner(&mut self, new_owner: PublicKey, version: u64) -> Result<(), DataError> {
+        if version != self.version + 1 {
+            return Err(DataError::InvalidSuccessor(self.version));
+        }
+        self.owners = new_owner;
+        self.version = version;
+        Ok(())
+    }
+
+    /// Change the owner without performing any validation.
+    pub fn change_owner_without_validation(&mut self, new_owner: PublicKey, version: u64) -> bool {
+        if version <= self.version {
+            return false;
+        }
+
+        self.owners = new_owner;
+        self.version = version;
+        true
+    }
+
+    pub fn is_action_allowed(&self, requester: PublicKey, action: Action) -> bool {
+        if self.owners == requester {
+            return true;
+        }
+        match self.permissions.get(&requester) {
+            Some(perms) => perms.clone().is_allowed(action),
+            None => false,
+        }
+    }
 }
 
-/// Subject of permissions
-#[derive(Hash, Eq, PartialEq, PartialOrd, Ord, Clone, Serialize, Deserialize)]
-pub enum User {
-    /// Permissions apply to a single public key.
-    Key(PublicKey),
-    /// Permissions apply to anyone.
-    Anyone,
+/// Implements functions which are COMMON for both the mutable data
```

Comment:
I think this comment should be replaced with something like
```suggestion
/// Implements functions for sequenced Mutable Data.
```